Ensuring compatibility with multi column and add-ons in general

Arthur-Milchior opened the pull request:
#### Description

Compatibility with Multi column add-on has been broken for some time.
The fact that both add-ons entirely redefine "loadNote" and replace all of "#fields" does not help.

![2020-03-24-085707_1916x1041_scrot](https://user-images.githubusercontent.com/357361/77401610-84065a00-6dad-11ea-80c4-0ae4059b0dee.png)

I did change to both add-ons so that:
Both add-on uses recent hooks, (a hook for note loading as been added recently, I'll love to use it when .24 is out). In particular load js file directly.
multi column redefine javascript's `setFields`, since it need to do a lot of change. This way, there is no need for event, the direct table is directly computed.
frozen field only do edition where the edition needs to be done. This allows it to avoid redefining the whole `setFields` function, and thus easily be compatible with other add-on.

And even better, for me, it makes both add-ons easily compatible with other add-ons dealing with editor, such as "resize image in editor" and "latex image in editor"


#### Checklist:

*Please replace the space inside the brackets with an **x** and fill out the ellipses if the following items apply:*

- [x] I've read and understood the [contribution guidelines](./CONTRIBUTING.md)
- [x] I've tested my changes against at least one of the following [Anki builds](https://apps.ankiweb.net/#download):
  - [x] Latest standard Anki 2.1 binary build [required for Anki-compatible 2.1 add-ons]
  - [ ] Latest alternative Anki 2.1 binary build
  - [ ] Latest Anki 2.0 binary build [required for Anki 2.0-compatible add-ons]
- [x] I've tested my changes on at least one of the following platforms:
  - [x] Linux, version:
  - [ ] Windows, version:
  - [ ] macOS, version: 
- [ ] My changes potentially affect non-desktop platforms, of which I've tested:
  - [ ] AnkiMobile, version:
  - [ ] AnkiDroid, version:
  - [ ] AnkiWeb

Arthur-Milchior committed:
Correct indentation

Arthur-Milchior committed:
Compatibility with multi column, and hooks

Compatibility between this add-on and multi column window has been
broken for some time. I did repair it in an add-on merging them, but
I'm not a fan of this solution.

Instead, here, I decided that instead of redefining entirely loadNote,
I wrap code around it. The code just add the frozen image.

In js, setFrozenFields is just in charge of adding the image before
the name. The image is .9em in height, so smaller than the text. It
even leads to more space in the field

Arthur-Milchior commented:
The related PR for multi column add-on is https://github.com/hssm/anki-addons/pull/32

Arthur-Milchior committed:
factorize fnames

Arthur-Milchior committed:
use webview_did_receive_js

Arthur-Milchior committed:
replace self by editor

glutanimate commented:
Thanks, Arthur! This looks very promising. Will look over it more thoroughly soon.

Arthur-Milchior commented:
FYI: HSSM and Damien gave me the "multi column editor" add-on. I just updated the add-on (for anki 22+). The change I sent to you works with and without the "multi column editor" add-on.

glutanimate commented:
Thanks for the update, Arthur. Glad to hear you're now maintaining "multi column editor"! Sorry for taking a bit on merging this. It's just that I've been super busy lately. Will look into it asap

Arthur-Milchior commented:
No hurry

ijgnd commented:
@glutanimate: If you could merge and release this soon this should avoid some problems. 

As far as I see once you have released this @Arthur-Milchior can remove his addon [Advanced note editor (Multi-column, Frozen fields, LaTeX's result in editor)](https://ankiweb.net/shared/info/2064123047) because then all of it's functionality should work from the original add-ons.

The latest version of Arthur's "Advanced note editor" was released before 2.1.22 so it doesn't use the 2.1.22 hooks and so it relies on overwrting central Editor methods which in turn breaks other add-ons (like an add-on of mine). It doesn't make sense for Arthur to update his temporary add-on which will be irrelevant once you've update frozen fields.

Thanks.

glutanimate commented on src/frozen_fields/main.py:
```diff
@@ -163,45 +94,37 @@ def frozenToggle(self, batch=False):
-    addHook("setupEditorButtons", onSetupButtons20)
-    Editor.bridge = wrap(Editor.bridge, onBridge, 'around')
-    Editor.loadNote = loadNote20
+addHook("setupEditorShortcuts", onSetupShortcuts)
```
Might be worthwhile to switch to `gui_hooks.editor_did_init_shortcuts`, while we're at it.

Arthur-Milchior replied:
done

glutanimate commented on src/frozen_fields/main.py:
```diff
@@ -36,114 +39,42 @@
     """Extends the js<->py bridge with our pycmd handler"""
 
+    if not isinstance(editor, Editor) or not isinstance(editor.parentWindow, AddCards):
+        return
```
This needs to return `handled`, otherwise entering any webview other than the editor causes:

```
TypeError: cannot unpack non-iterable NoneType object
```

Arthur-Milchior replied:
done. Thanks.

glutanimate reviewed (changes requested):
First of all: Thanks again for all the work you've put into resolving these conflicts, Arthur! It's fantastic to finally have a potential solution to an issue that has been plaguing these add-ons for so long.

I added a couple of comments, but the only one that's truly important is fixing `onBridge`.

Outside of that your PR seems to work well so far. Nicely done!

The only remaining issues I see are on the UX side. Because Frozen Fields has a pretty huge userbase, most of which have been using the add-on for years now, I'm very hesitant about introducing major changes to how users interact with the add-on. The two key issues I see with these changes at the moment are:

1. When using the hotkeys to toggle fields, there is a noticeable shifting around / flickering of the field names which could be jarring for users:

*old*

![old mp4](https://user-images.githubusercontent.com/5459332/80388920-c0484100-88aa-11ea-9bac-77a31c4211bc.gif)

*new*

![new mp4](https://user-images.githubusercontent.com/5459332/80388915-bfafaa80-88aa-11ea-96c6-09955e2bb1c9.gif)

This is one of the issues #5 also had. If you could find a way to address that, I'd really appreciate it.

2. Although I mostly use the add-on through its key-bindings (which I'm sure you guys mostly do, too), I assume there's a significant portion of users that interact with it through their mouse (potentially even the majority of users). For these users, the drastic reduction of clickable space and shift to a different position would likely be an issue.

    Would it be possible for you to maintain the original position & size of the icon without running into any conflicts with your other add-ons?

If we can get both of these out of the way (and the exception), I'd be more than happy to merge this and push it to AnkiWeb.

(I would likely check if I can find a way to make the changes backwards-compatible with <2.1.22 first as that's where the overwhelming portion of users still is; but I also wouldn't be opposed to creating a new branch on AnkiWeb if need be (although I hate doing that because it incurs a lot of additional maintenance & support effort).)

@ijgnd I appreciate the reminder, as I'd lost track of this review! I can definitely see the need for resolving these series of conflicts, and I hope we can get them out of the way with this once and for all –  I've been bitten by them myself way too many times.

However, for add-ons with as big of a userbase as this, please understand that I have to be very careful about the changes I introduce. This includes considerations about Anki version adoption, adding new features vs introducing regressions, and user response.

Arthur-Milchior committed:
remove one if anki21

Arthur-Milchior committed:
onBridge assumed 21

Arthur-Milchior committed:
frozen toggle assume 21

Arthur-Milchior committed:
remove unused loadNote20

Arthur-Milchior committed:
remove unused js_code_20

Arthur-Milchior committed:
remove unused onSetupButtons20

Arthur-Milchior committed:
remove unused js20.js

Arthur-Milchior committed:
remove 21 in names

Arthur-Milchior commented:
Does not work anymore. I don't get why. I'm looking at it

Arthur-Milchior committed:
js imported through webview_will_set_content

Arthur-Milchior committed:
editor_did_init_shortcuts

Arthur-Milchior committed:
set back old star position

Arthur-Milchior commented:
Corrected.
I put back the star at its old place.

Many add-ons want to add a button to some field. As an example, "multi column window" have a button which allow to switch between field having full width and small field. Cleaning html I sent you recently have a button which allow to clean one particular field. I find that having the button above the field is far more natural than putting the cursor in the field and then clicking the button on top of the editor.
I was trying to ensure that frozen field was consistent with all other add-ons I created which add feature on fields themselves. And also to create a pattern that other may follow, that we may had all buttons we want near the name, where there is a lot of place.
I admit that it forces me to have a small button in order not to loose space

I would be okay with creating an option in the add-on config to switch between both style.

Arthur-Milchior committed:
use only $img in onFrozen

Arthur-Milchior committed:
remove elem from onFrozen

Arthur-Milchior committed:
remove toggle of stars

Arthur-Milchior commented:
I pushed again to remove the flickering. This actually allow to simplify the code and totally avoid to reload the note. It sends a message to JS to tell it to change the star, and js sends the message to python to tell to change the status of "sticky"

Arthur-Milchior committed:
add zipping

Arthur-Milchior committed:
before instead of prepend

With multi column, it ensure that the star is added before the field
and not at the start of the line

Arthur-Milchior committed:
allowing multi column editor to have full size field

Arthur-Milchior commented:
I just pushed a correction to the last problem I had: having fields which takes all columns

Arthur-Milchior committed:
extend the name too

Otherwise if the name is too small, it extend the column size uselessly

Arthur-Milchior commented:
![2020-05-03-021010_956x1041_scrot](https://user-images.githubusercontent.com/357361/80895279-5d5c0d00-8ce3-11ea-8906-589ef93751e8.png)
I don't know how to make gif, so I can't as easily show that there is no flickering

Arthur-Milchior committed:
split img in two variables

Arthur-Milchior committed:
add an id to images' td

Arthur-Milchior committed:
change image if changed

Arthur-Milchior commented:
While using anki, I realized an unexpected problem. Sometime, anki asks the note to be loaded twice quickly, which does not let callback enough time to be executed. So the code was executed twice. This was not a problem before because the code erased everything and replaced by a new table. Now that the code modify the current table, this led to errors. I corrected it by ensuring that if Frozen-fields' code was already executed, it ensure that the star has the correct value instead of adding a new star
